src/main/java/com/google/devtools/build/lib/vfs/OutputService.java

```diff
@@ -110,14 +110,19 @@ default RemoteArtifactChecker getRemoteArtifactChecker() {
   /**
    * Start the build.
    *
+   * @param outputPath absolute path pointing to the output directory
    * @param buildId the UUID build identifier
    * @param finalizeActions whether this build is finalizing actions so that the output service can
    *     track output tree modifications
    * @return a ModifiedFileSet of changed output files.
    * @throws BuildFailedException if build preparation failed
    * @throws InterruptedException
    */
-  ModifiedFileSet startBuild(EventHandler eventHandler, UUID buildId, boolean finalizeActions)
+  ModifiedFileSet startBuild(
+      Path outputPath,
```

Comment:
Some of the other methods of OutputService already seem to take `PathFragment execRootFragment, String relativeOutputPath`. I thought about using that here as well, but that also means that I'd need to pass in a `FileSystem` separately. All of that seemed a bit too convoluted, which is why I used a simple `Path` for the entire output path instead.

Reply:
Hmmm... I just tried doing that, but that's not entirely trivial. To construct the output path, we not only need to know the pathnames, but also the workspace name. If I make a change along these lines:
```
--- a/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java
@@ -141,6 +141,7 @@ public final class RemoteModule extends BlazeModule {
   @Nullable private RemoteActionContextProvider actionContextProvider;
   @Nullable private RemoteActionInputFetcher actionInputFetcher;
   @Nullable private RemoteOptions remoteOptions;
+  @Nullable private CommandEnvironment env;
   @Nullable private RemoteOutputService remoteOutputService;
   @Nullable private TempPathGenerator tempPathGenerator;
   @Nullable private BlockWaitingModule blockWaitingModule;
@@ -313,6 +314,7 @@ public final class RemoteModule extends BlazeModule {
     }
 
     this.remoteOptions = remoteOptions;
+    this.env = env;
 
     AuthAndTLSOptions authAndTlsOptions = env.getOptions().getOptions(AuthAndTLSOptions.class);
     DigestHashFunction hashFn = env.getRuntime().getFileSystem().getDigestFunction();
@@ -914,6 +916,7 @@ public final class RemoteModule extends BlazeModule {
     actionContextProvider = null;
     actionInputFetcher = null;
     remoteOptions = null;
+    env = null;
     remoteOutputService = null;
     tempPathGenerator = null;
     rpcLogFile = null;
@@ -1067,7 +1070,8 @@ public final class RemoteModule extends BlazeModule {
   public OutputService getOutputService() {
     Preconditions.checkState(remoteOutputService == null, "remoteOutputService must be null");
     if (actionContextProvider.getRemoteCache() != null) {
-      remoteOutputService = new RemoteOutputService();
+      remoteOutputService = new RemoteOutputService(
+          env.getDirectories().getOutputPath(env.getWorkspaceName()));
     }
     return remoteOutputService;
   }
```
I see tests fail with this backtrace:
```
java.lang.NullPointerException
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:889)
        at com.google.devtools.build.lib.runtime.CommandEnvironment.getWorkspaceName(CommandEnvironment.java:561)
        at com.google.devtools.build.lib.remote.RemoteModule.getOutputService(RemoteModule.java:1074)
        at com.google.devtools.build.lib.runtime.CommandEnvironment.beforeCommand(CommandEnvironment.java:766)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:544)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:244)
        at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:550)
        at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$1(GrpcServerImpl.java:621)
        at io.grpc.Context$1.run(Context.java:566)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

So CommandEnvironment.beforeCommand() depends on RemoteModule.getOutputService(). But now RemoteModule.getOutputService() will start to depend on CommandEnvironment being fully initialized.

Any idea on how I can break that cycle?

Reply:
I was initially running into cyclic dependencies, because I was trying to construct the `outputPath` inside of `RemoteModule.getOutputService()`. Unfortunately, `CommandEnvironment.beforeCommand()` itself already depends on that function. But if we pass through the entire `CommandEnvironment` into `RemoteOutputService`, we can make it work. Thanks!